### PR TITLE
Improve handling of partition label type for multipath disks

### LINKED ISSUE #2234: Multipath disk restored with MBR instead of GPT
* ReaR version ("/usr/sbin/rear -V"):

      # ./usr/sbin/rear -V
      Relax-and-Recover 2.5 / Git

  Git revision a252ca20b5dc9f34f6c8fffb61e055141ca37669 (current HEAD).

* OS version ("cat /etc/rear/os.conf" or "lsb_release -a" or "cat /etc/os-release"):

      # cat etc/rear/os.conf 
      OS_VENDOR=SUSE_LINUX
      OS_VERSION=12.2

* ReaR configuration files ("cat /etc/rear/site.conf" and/or "cat /etc/rear/local.conf"):

      # cat etc/rear/local.conf
      BACKUP=NETFS
      BACKUP_OPTIONS="nfsvers=4,nolock"
      BACKUP_URL=nfs://192.168.122.1/srv/nfs/multipath-test
      REQUIRED_PROGS=( "${REQUIRED_PROGS[@]}" snapper chattr lsattr )
      COPY_AS_IS=( "${COPY_AS_IS[@]}" /usr/lib/snapper/installation-helper /etc/snapper/config-templates/default )
      BACKUP_PROG_INCLUDE=( /var/cache /var/lib/mailman /var/tmp /var/lib/pgsql /usr/local /opt /var/lib/libvirt/images /boot/grub2/i386-pc /var/opt /srv /boot/grub2/x86_64-efi /var/lib/mariadb /var/spool /var/lib/mysql /tmp /home /var/log /var/lib/named /var/lib/machines )
      AUTOEXCLUDE_DISKS=n
      AUTOEXCLUDE_MULTIPATH=n

* Hardware (PC or PowerNV BareMetal or ARM) or virtual machine (KVM guest or PoverVM LPAR):

  KVM guest.

* System architecture (x86 compatible or PPC64/PPC64LE or what exact ARM device):

  x86-64.

* Firmware (BIOS or UEFI or Open Firmware) and bootloader (GRUB or ELILO or Petitboot):

  UEFI + GRUB.

* Storage (local disk or SSD) and/or SAN (FC or iSCSI or FCoE) and/or multipath (DM or NVMe):

  DM multipath disk with GPT.

* Description of the issue (ideally so that others can reproduce it):

  When recording information about a multipath disk, ReaR currently does not store information about its partition label type (file `280_multipath_layout.sh`). Example multipath line:

      multipath /dev/mapper/0QEMU_QEMU_HARDDISK_0001 2147483648 /dev/sda,/dev/sdb

  During recovery, `create_multipath()` (`210_load_multipath.sh`) is used to process a single multipath definition. The function primarily calls `create_partitions()` (`100_include_partition_code.sh`) to create a label and partitions on the device. However, no label type is passed to `create_partitions()` which causes that the code tries to automatically detect the label type. This heuristics works as follows:

          ### List partition types/names to detect disk label type.
          local -a names=()
          local part size pstart name junk
          while read part disk size pstart name junk ; do
              names=( "${names[@]}" $name )
              case $name in
                  (primary|extended|logical)
                      if [[ -z "$label" ]] ; then
                          Log "Disk label for $device detected as msdos."
                          label="msdos"
                      fi
                      ;;
              esac
          done < <( grep "^part $device " "$LAYOUT_FILE" )

  The logic goes wrong if a multipath disk with GPT has a partition with a name set to exactly "primary", "extended" or "logical". In such a case, ReaR creates the MBR label on the restored disk instead of GPT.

  Note: Restoring normal disks does not rely on this heuristic because the partition label type is explicitly stored for them.

* Workaround, if any:

  The problem can be avoided by using the gdisk or parted utility and renaming all GPT partitions with names "primary", "extended" or "logical" to something different.

### PR DESCRIPTION
**Pull request by petrpavlu**
##### Pull Request Details:

* Type: **Bug Fix** / **New Feature** / **Enhancement** / **Other?**

  **Bug Fix.**

* Impact: **Low** / **Normal** / **High** / **Critical** / **Urgent**

  **Normal.**

* Reference to related issue (URL):

  https://github.com/rear/rear/issues/2234.

* How was this pull request tested?

  Locally tested under KVM. Multipath configuration:

      # multipath -ll
      0QEMU_QEMU_HARDDISK_0001 dm-0 QEMU,QEMU HARDDISK
      size=2.0G features='1 retain_attached_hw_handler' hwhandler='0' wp=rw
      |-+- policy='service-time 0' prio=1 status=active
      | `- 2:0:0:1 sdb 8:16 active ready running
      `-+- policy='service-time 0' prio=1 status=enabled
        `- 2:0:0:2 sda 8:0  active ready running

      # parted /dev/mapper/0QEMU_QEMU_HARDDISK_0001 print
      Model: Linux device-mapper (multipath) (dm)
      Disk /dev/mapper/0QEMU_QEMU_HARDDISK_0001: 2147MB
      Sector size (logical/physical): 512B/512B
      Partition Table: gpt
      Disk Flags: 
      
      Number  Start   End     Size    File system  Name     Flags
       1      1049kB  211MB   210MB   ext4         primary
       2      211MB   2147MB  1937MB  xfs          primary

* Brief description of the changes in this pull request:

  When recording information about a multipath disk, ReaR did not store information about its partition label type (`280_multipath_layout.sh`). The recovery code `create_multipath()` (`210_load_multipath.sh`) -> `create_partitions()` (`100_include_partition_code.sh`) then tried to automatically detect the label type using a heuristic that depends on GPT partition names. The logic would incorrectly detect the device as having the MBR label type instead of GPT if one of the partition names was exactly "primary", "extended" or "logical".

  The patch fixes the problem by explicitly storing the partition label type for multipath devices as is already done for normal disks. The restore logic is accordingly updated to utilize this information.

  Fixes #2234.

### CODE REVIEW

**Comment by jsmeix** [comment]
@schabrolles 
could you please review it because I know basically nothing about multipath.

File: usr/share/rear/layout/save/GNU/Linux/280_multipath_layout.sh [hunk 1]
```diff
@@ -23,7 +23,11 @@ while read dm_name junk ; do
         slaves="$slaves$(get_device_name ${slave##*/}),"
     done
 
-    echo "multipath /dev/mapper/$dm_name $dm_size ${slaves%,}" >> $DISKLAYOUT_FILE
+    dm_disktype=$(parted -s $dev_name print | grep -E "Partition Table|Disk label" | cut -d ":" -f "2" | tr -d " ")
+
+    echo "# Multipath /dev/mapper/$dm_name"
+    echo "# Format: multipath <devname> <size(bytes)> <partition label type> <slaves>"
```
Comment (jsmeix) [inline usr/share/rear/layout/save/GNU/Linux/280_multipath_layout.sh]: Optionally while you are at it:

Could you also document the multipath syntax
in the "Disk layout file syntax" section in doc/user-guide/06-layout-configuration.adoc
https://github.com/rear/rear/blob/master/doc/user-guide/06-layout-configuration.adoc

Comment (setupji) [reply usr/share/rear/layout/save/GNU/Linux/280_multipath_layout.sh]: Added in the updated patch.

**Comment by jsmeix** [comment]
@schabrolles 
if you do not object, I would like to merge it tomorrow.

**Pull request merged**